### PR TITLE
insights: update gql to allow control over number of series samples

### PR DESCRIPTION
**Pull request by leonore**
closes https://github.com/sourcegraph/sourcegraph/issues/45750

open https://github.com/sourcegraph/sourcegraph/issues/46283 as follow up, frontend changes still need to be done

## Test plan

Added unit tests
Manually tested:

```gql
query GetInsightView($id: ID, $filters: InsightViewFiltersInput, $seriesDisplayOptions: SeriesDisplayOptionsInput) {
  insightViews(
    id: $id
    filters: $filters
    seriesDisplayOptions: $seriesDisplayOptions
  ) {
    nodes {
      ...InsightDataNode
      __typename
    }
    __typename
  }
}

fragment InsightDataNode on InsightView {
  id
  dataSeries {
    ...InsightDataSeries
    __typename
  }
  __typename
}

fragment InsightDataSeries on InsightsSeries {
  seriesId
  label
  points {
    dateTime
    value
    diffQuery
    __typename
  }
  __typename
}
```

variables
```json
{"id":"aW5zaWdodF92aWV3OiIySjA3YTE4bGZTZEZzenYxOXp4QjhrZWQ5RVki","filters":{"includeRepoRegex":"","excludeRepoRegex":"","searchContexts":[""]},"seriesDisplayOptions":{"limit":20,"sortOptions":{"direction":"DESC","mode":"RESULT_COUNT"}, "numSamples": 1}}
```

result
```json
{
  "data": {
    "insightViews": {
      "nodes": [
        {
          "id": "aW5zaWdodF92aWV3OiIySjA3YTE4bGZTZEZzenYxOXp4QjhrZWQ5RVki",
          "dataSeries": [
            {
              "seriesId": "2J07a7NrF1cNQjEGfFF8XU2ygmR",
              "label": "I hate",
              "points": [
                {
                  "dateTime": "2023-01-10T13:44:42Z",
                  "value": 2,
                  "diffQuery": "repo:leonore before:2023-01-10T13:44:42Z type:diff I hate",
                  "__typename": "InsightDataPoint"
                }
              ],
              "__typename": "InsightsSeries"
            }
          ],
          "__typename": "InsightView"
        }
      ],
      "__typename": "InsightViewConnection"
    }
  }
}
```

### CODE REVIEW

File: cmd/frontend/graphqlbackend/insights.graphql [hunk 1]
```diff
@@ -671,6 +671,10 @@ type SeriesDisplayOptions {
     Max number of series to return.
     """
     limit: Int
+    """
+    Max number of samples to return for all the series on the view.
+    """
+    numSamples: Int
```
Comment (chwarwick) [inline cmd/frontend/graphqlbackend/insights.graphql:677]: It feels wrong to include this is the options of the series if this isn't persisted and is just returning whatever was sent in the input options.

**Review by chwarwick — APPROVED** [review]
Is the plan to persist this value in the future?   I think there 2 potential impacts from this.
- hourly/daily/weekly insights could start showing fewer points by default - I think this is ok
- Monthly insights start showing more points - I also think this is ok

I do think we should remove it from the resolver if it is just returning user input.

File: cmd/frontend/graphqlbackend/insights.graphql (hunk 1 above)
Comment (coury-clark) [reply cmd/frontend/graphqlbackend/insights.graphql:677]: It does allow us to "soft validate" by replacing invalid (too big) values with a max. I'm not strongly opinionated either way though.

Comment (leonore) [reply cmd/frontend/graphqlbackend/insights.graphql:677]: I can remove it until I work on this issue if you'd rather: https://github.com/sourcegraph/sourcegraph/issues/46283

**Comment by leonore** [comment]
> Is the plan to persist this value in the future? I think there 2 potential impacts from this.

yes that is the plan (https://github.com/sourcegraph/sourcegraph/issues/46283) 

it's just a lot of boilerplate to implement that functionality, so I didn't want to detract from the actual logic being implemented here

File: enterprise/internal/insights/store/store.go [hunk 2]
```diff
@@ -682,6 +688,27 @@ func (s *Store) GetInsightSeriesRecordingTimes(ctx context.Context, id int, from
...
 }
 
+func (s *Store) GetOffsetNRecordingTime(ctx context.Context, seriesId, n int, excludeSnapshot bool) (time.Time, error) {
+	preds := []*sqlf.Query{sqlf.Sprintf("insight_series_id = %s", seriesId)}
+	if excludeSnapshot {
+		preds = append(preds, sqlf.Sprintf("snapshot is false"))
+	}
+
+	var tempTime time.Time
+	oldestTime, got, err := basestore.ScanFirstTime(s.Query(ctx, sqlf.Sprintf(getOffsetNRecordingTimeSql, sqlf.Join(preds, "and"), n)))
```
[the remark this answers is not in the document]
Comment (leonore) [reply enterprise/internal/insights/store/store.go:698]: all series would have gotten updated at the tail and at the head. but that wouldn't change the fact that number of recording_times will always be = number of series points, so we can safely fetch the series points based on this timestamp

File: cmd/frontend/graphqlbackend/insights.graphql (hunk 1 above)
Comment (leonore) [reply cmd/frontend/graphqlbackend/insights.graphql:677]: leaving this in, added soft-validation (max 90 samples). that issue linked will take care on allowing to actually set a default number of samples (which will need frontend work to actually work...)

**Pull request merged**